tests/ui/needless_cow.rs

```diff
@@ -0,0 +1,71 @@
+#![allow(
+    clippy::borrow_deref_ref,
+    clippy::collapsible_else_if,
+    clippy::deref_addrof,
+    clippy::if_same_then_else
+)]
+#![warn(clippy::needless_cow)]
+
+use std::borrow::Cow;
+
+// FIXME: Should not lint.
+fn passed_to_fn() {
+    fn helper(_: Cow<'static, str>) {}
+
+    let x = if true { Cow::Borrowed("") } else { Cow::Borrowed("") };
+    let y = x;
+
+    helper(y);
+}
+
+// FIXME: Should not lint.
+fn passed_to_fn_mut_ref() {
+    fn helper(_: &mut Cow<'static, str>) {}
+
+    let x = if true { Cow::Borrowed("") } else { Cow::Borrowed("") };
+    let mut y = x;
+
+    helper(&mut y);
+}
+
+// FIXME: Should not lint.
+fn returning() -> Cow<'static, str> {
+    if true { Cow::Borrowed("") } else { Cow::Borrowed("") }
+}
```

Comment:
We can probably do another analysis pass for these, but I'm unsure if that's the best way to fix this. The issue is that it doesn't *have* to refer to the local initialized by `Borrowed`, it could be something like `_0 = _2` where `_2 = _3` which makes it a bit annoying. There's probably an analysis pass that gives you which locals it "depends" on